### PR TITLE
ENH: override isf for Burr distribution

### PR DESCRIPTION
**Pull request by OmarManzoor**
#### Reference issue

Towards: gh-17832

#### What does this implement/fix?

- Overrides and adds the inverse survival function of the Burr distribution in order to improve its precision
- Adds some tests.

#### Additional information

I still have to add the reference distribution if required.
CC: @mdhaber

### CODE REVIEW

**Comment by OmarManzoor** [comment]
Figure_1

<Details>

```python
from scipy.stats import burr
from scipy.special import expm1, xlog1py
import numpy as np
import matplotlib.pyplot as plt
from mpmath import mp
mp.dps = 200

def mpmath_isf(q, c, d):
    q = mp.mpf(q)
    c = mp.mpf(c)
    d = mp.mpf(d)

    q = mp.one - q
    value = (q**(-mp.one/d) - 1)**(-mp.one/c)
    return float(value)


def _isf(q, c, d):
    _p = xlog1py(-1.0/d, -q)
    return expm1(_p)**(-1.0/c)

def plot_isf():
    c, d = 5.0, 3.0
    q = np.logspace(-40, -1, 200)
    mp_values = np.array([mpmath_isf(_q, c, d) for _q in q], np.float64)
    plt.semilogx(q, burr.isf(q, c, d), label="main", ls="dashed")
    plt.semilogx(q, _isf(q, c, d), label="pr", ls="dashdot")
    plt.semilogx(q, mp_values, label="mpmath", ls="dotted")
    plt.legend()
    plt.title("Burr inverse survival function")
    plt.show()
```

</Details>

**Comment by mdhaber** [comment]
All three lines on the plot appear to overlap perfectly. Can you adjust the plot to show the improvement? For instance:
- you can use the y axis to plot the relative error rather than the value of the result
- you can use log scale on the y axis to highlight the magnitude of the error

**Comment by OmarManzoor** [comment]
> All three lines on the plot appear to overlap perfectly. Can you adjust the plot to show the improvement? For instance:
> 
> * you can use the y axis to plot the relative error rather than the value of the result
> * you can use log scale on the y axis to highlight the magnitude of the error

Actually the one for main stops way before. The color for main is just not clear.
Figure_1

**Comment by mdhaber** [comment]
I see. Makes sense.

File: scipy/stats/tests/test_distributions.py [hunk 1]
```diff
@@ -7109,6 +7109,16 @@ def test_burr_nan_mean_var_9544(self):
         assert_(np.isfinite(e3))
         assert_(np.isfinite(e4))
 
+    def test_burr_isf(self):
+        # reference values were computed via the reference distribution, e.g.
+        # mp.dps = 100
+        # Burr(c=c, d=d).sf(q).
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:7115]: If the reference values were created with `sf`, is the `_ppf` method of the reference distribution needed?

**Review by mdhaber — APPROVED** [review]
Looks accurate and it avoids subtractive cancelation of the argument of `isf`.

Just a question about the reference values.

File: scipy/stats/tests/test_distributions.py (hunk 1 above)
Comment (OmarManzoor) [reply scipy/stats/tests/test_distributions.py:7115]: I tried with just the pdf and it returned an error related to not being able to find a root for the given tolerance. Then I just added the cdf and it resulted in zero division errors. Thus I tried added the ppf and it seemd to returned the desired correct values.

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:7115]: That's fine; it's mostly about the comment. It says you used the `sf` method here. Did you use the `sf` method, or did you use the `ppf`/`isf` method?

Comment (OmarManzoor) [reply scipy/stats/tests/test_distributions.py]: Fixed the comment. Thanks for pointing it out.

File: scipy/stats/tests/test_distributions.py [hunk 2]
```diff
@@ -7109,6 +7109,16 @@ def test_burr_nan_mean_var_9544(self):
         assert_(np.isfinite(e3))
         assert_(np.isfinite(e4))
 
+    def test_burr_isf(self):
+        # reference values were computed via the reference distribution, e.g.
+        # mp.dps = 100
+        # Burr(c=c, d=d).isf(q).
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:7115]:
```suggestion
        # Burr(c=5, d=3).isf([0.1, 1e-10, 1e-20, 1e-40])
```

**Review by mdhaber — APPROVED** [review]
This looks accurate, and I verified the reference values. Note that even without defining the `_ppf` method, they can be produced by providing a `guess`, e.g. `Burr(c=5, d=3).isf(0.1, guess=2)`. However, the it is more convenient to be able to use `isf` without a guess, and the `ppf` implementation is easy to verify against the one we already had in SciPy, so I'll leave it.

**Pull request merged**